sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,363 @@
+in Kubernetes. In order to make a reliable extension mechanism for admission,
+we must reconsider the current chain to produce understandable extensions.
+
+We are proposing a series of changes to admission control in an attempt to make
+it less error-prone, easier to configure, and more flexible for admission plugin
+developers.
+
+## Requirements for extensible admission
+Ordering the admission chain is critical.  One mutating admission plugin can invalidate
+the results of other admission plugins.  For instance, the `LimitVerifier` would
+invalidate results for the `ResourceQuota` plugin.  More generally, object mutations
+must happen before object validations.  In addition, compiled validating admission plugins
+must run after mutating admission webhooks.  Splitting the chain into mutators and validators
+will resolve the coarse grained ordering problem and would require two different webhook APIs
+since the `Status` would be different for both now and we can reasonably expect different
+evolutions.  It would also force changes in the registration APIs since ordering matters
+for individual mutators, but ordering does not matter for validators.  Having fewer items to 
+order makes the registration API easier to reason about for maintainability.
+
+Initializers as implemented today attempt to push new API validation requirements into the 
+strategy and new requirements onto the admission plugins handling updates and creates.  The
+former is unnecessary and the latter doesn't work.
+
+API validation on uninitialized objects can reasonably follow the rules for validation on create
+plus `ValidateObjectMetaUpdate` (to prevent name and UID changes and the like).
+Since the object hasn't been "ready", there is no reason to add the validation rules required on
+updates since restrictions like field immutability don't logically apply and would restrict the power
+of admission extensions to the point of unusability.  Using the results for creation validation on 
+uninitialized objects and objects transitioning from uninitialized to initialized lets the
+strategy author focus on their API needs.
+
+Admission validation is also needed on resources being initialized.  During initialization, there
+may be some admission validation required, but it isn't normal API update admission.  The update
+admission mutators shouldn't run and since the subject (user.Info) on the admission attributes is
+NOT the original user, normal update admission mutators and validators are likely to make improper
+choices.  When the object transitions from uninitialized to initialized state, the original object
+creator should be artificially injected into the admission attributes and the create admission 
+validators should run.  This allows mechanisms like PSP to function properly during their validation
+phase.  Such injection requires persisting the original user.Info (probably encrypted) into the 
+object.
```

Comment:
> How does PSP handle normal UPDATE initiated by controllers? Does it have the same problem?

Normal updates don't get to mutate the important bits of the pod spec.

Reply:
> Normal updates don't get to mutate the important bits of the pod spec.

Yes, I understand that. Let me ask the question more specifically: the PSP AC [handles](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L91) UPDATE. And in the beginning of its `Admit()` method, it [finds](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L134) the matched psp based on the user info. If a controller updates the object, does that mean the PSP controller will find the policy for the controller rather than for the original creator?

Also, assuming an initializer adds a side-car container to the pod, should that container subject to the original creator's psp, or the initializer's psp? I think it should be the later.

Reply:
I take that back, the image is able to be updated, which is why PSP cares about update at all. Also, I think there are some pod annotations being used for features like App armor that are mutable by virtue of being  annotations

Reply:
> Are there examples other than the PSP?

The fact that we needed it so quickly indicates that the use-case has a lot of value.


> Also, assuming an initializer adds a side-car container to the pod, should that container subject to the original creator's psp, or the initializer's psp? I think it should be the later.

Post-validation mutation (which is what this is describing) would be possible to describe, but it seems really risky.  If it were expressed in our current admission chain, the second mutation (adding a side-car container than the requestor couldn't request), would not be validated at all.

I don't think it would be a good idea to add that behavior in a rush.  Do we have any current admission plugins in our chain trying to do this?  I don't know of any.

Reply:
Not any in-tree admission plugins. But out-of-tree, adding a side-car container is a common use case of initializers, I know a couple of google projects that use initializers this way.

That aside, I don't think recording the original user info is required. Pods are created in two ways: 1. by controllers (e.g., by the replicaset controller) 2. manually by end-user.
For case 1, like what's suggested in the PSP [doc](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#working-with-rbac), PSP should be bound to service account presented in the podTemplate. The original creator's identity is not important. Also, users shouldn't give initializers superuser permissions, like they shouldn't grant such permissions to controllers.
For case 2, I think first we don't encourage user to create pod manually, and second users can always use the service account to workaround.

cc @tallclair

Reply:
> For case 2, I think first we don't encourage user to create pod manually, and second users can always use the service account to workaround.

Can we just deprecate this approach? I think it makes the PSP usage harder to reason about, and in production users shouldn't really be creating pods directly (outside of debugging, maybe). In the long term, maybe we can come up with a mechanism to attribute pod creation & update back to the original user (although I'm not entirely sure that's desireable).

> Also, I think there are some pod annotations being used for features like App armor that are mutable by virtue of being annotations

Yeah, although we're trying to move away from that.

Reply:
Knowing the original user is a key piece of admission used by PSP, NodeRestriction, GC permission checks, and eventratelimiting.  Eliminating the power for dynamic admission hooks makes it impossible to build important classes of admission plugins.  The workaround of SAs only works for one of those checks.

Reply:
I think recording the original creator will be useful. Actually its usefulness goes beyond the dynamic admission. I'm trying to argue that it's not an urgent goal for beta, because dynamic admission mechanisms can support existing plugins without this feature.

My assumption regarding when the validation webhooks are run:
* Currently they are exercised to validate the original CREATE, and the intermediate initialization UPDATEs.
* In the future, for performance sake, webhooks can skip the intermediate initialization UPDATEs. They are exercised to validate the original CREATE, and the last UPDATE that completes the initialization.

If so, then for the admission plugins in question:
* NodeRestriction: it only handles mirror pods, which are exempted from initialization, so no initializer will mutate the pods that NodeRestrction cares about.

* GC permission checks: the plugin cares about if the modifier of the ownerReference has "delete" permission of the dependent object, or the "finalize" permission of the owner object. If an initializer modifies the ownerReference, I think it's reasonable to check if the initializer controller has the permissions, rather than checking the permissions of the original creator.

* eventratelimiting: the original CREATE will be validated, and will be rejected if the user creates too many events. Further, I think we should ban initializing events, otherwise there will be performance issue given the huge number of events.

* PSP: should use ServiceAccount instead ([details](https://github.com/kubernetes/community/pull/1137#discussion_r142241079))

To make these plugins dynamic, they can be adapted as webhooks. The apiserver sends the `User.Info()` to the webhooks.

Taking a step back, recording the original user is backwards compatible, we can add that after beta.

Reply:
> I think recording the original creator will be useful. Actually its usefulness goes beyond the dynamic admission. I'm trying to argue that it's not an urgent goal for beta, because dynamic admission mechanisms can support existing plugins without this feature.

If you change PSP you will break use-cases like an infrastructure component that wants to escalate a user by using a privileged identity for a specifically built pod (we know this use-case exists since we use it openshift).  

Attempting to run the current chain with PSP, without having this information can result in security exposure.  That breaks the current admission chain and that cannot be allowed for beta.

Reply:
> If you change PSP you will break use-cases like an infrastructure component that wants to escalate a user by using a privileged identity for a specifically built pod (we know this use-case exists since we use it openshift).

For such specific pod, can it just skip initialization?

Reply:
> For such a specific pod, can it just skip initialization?

The controller itself is still subject to PSP and should still be subject to PSP.  It is simply more powerful than the user, not all powerful.

Reply:
Do you mean it should be subject to initializers? Is it possible to point me to the code? It might be easier to understand the problem.

Reply:
> Knowing the original user is a key piece of admission used by PSP, NodeRestriction, GC permission checks, and eventratelimiting. Eliminating the power for dynamic admission hooks makes it impossible to build important classes of admission plugins. The workaround of SAs only works for one of those checks.

Are you proposing that updates be checked against PSP's accessible by the original creator's identity? I think that opens up an attack pattern of swapping out the image on an existing privileged pod...

I think the PSP needs to be checked with the updater's identity (+pod SA identity), i.e. no changes to how it works today. I'm not sure I see any way around it, but would love to be proven wrong.

Reply:
Maybe the PSP plugin can whitelist the fields it doesn't care, and allow initializers to change them w/o reevaluating the entire pod. And for the sensitive fields, initializers will need proper security permission to update them.

Reply:
Just thought of a similar problem: today the garbage collector controller updates pods to change ownerRefs . How does the PSP plugin handle that? Does it give the GC controller super permission?

Reply:
> today the garbage collector controller updates pods to change ownerRefs . How does the PSP plugin handle that? 

Funny you should ask: https://github.com/kubernetes/kubernetes/commit/32e5147af94f5becd359a71e3e1f421e9b8a721c

> Does it give the GC controller super permission?

Definitely not.

Reply:
That looks good. @liggitt can we do similar things for initializers? Essentially what I proposed in the earlier [comment](https://github.com/kubernetes/community/pull/1137#discussion_r143064604).